Add ErrorHandling Middleware

markstory opened the pull request:
Having middleware that can render errors reduces our reliance on global exception handlers and allows middleware to be wrapped around the error handling allowing things like CORS headers to be set on error pages.

Refs #6960 #7683

markstory committed:
Add ErrorHandling Middleware

Having middleware that can render errors reduces our reliance on global
exception handlers and allows middleware to be wrapped around the error
handling allowing things like CORS headers to be set on error pages.

lorenzo commented on src/Http/Middleware/ErrorHandlerMiddleware.php:
```diff
@@ -0,0 +1,101 @@
+        try {
+            $response = $renderer->render();
+            return ResponseTransformer::toPsr($response);
+        } catch (Exception $e) {
```
Can you alias this exception? I found a bit confusing at first

markstory replied:
Sure thing.

lorenzo commented on src/Http/Middleware/ErrorHandlerMiddleware.php:
```diff
@@ -0,0 +1,101 @@
+            );
+            trigger_error($message, E_USER_ERROR);
+        }
+        return $response;
```
Shouldn't we delegate somehow the response handling of this fatal error case? Ideally we still want to render a 500 error page, but instead we are letting the rest of the middleware stack to finish without an error code

markstory replied:
We could augment the response status code to a 50x and set a text response in addition to triggering an error.

markstory replied:
I've updated the code to update the response and log an error instead of triggering a fatal error. Let me know what you think.

markstory committed:
Instead of triggering a fatal error, modify the response.

Modify the response with a text response and log an error when rendering
an error page.

lorenzo commented:
Great job @markstory :)
